### PR TITLE
Add LSP support for Lua (EmmyLua-LS-all).

### PR DESCRIPTION
**Pull request by sunlin7**
Add the LSP support for Lua (EmmyLua-LS-all), and it's easy to extend with other Lua LSP backend.

### CODE REVIEW

**Comment by sunlin7** [comment]
Hi @smile13241324 , This PR include the last modifications according your comment https://github.com/syl20bnr/spacemacs/pull/12927#pullrequestreview-322548132. 
Please review it. Thank you so much.

File: layers/+lang/lua/funcs.el [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      (setq lsp-clients-emmy-lua-jar-path (expand-file-name lua-lsp-emmy-jar-path)))
+    (setq lsp-enable-file-watchers lua-lsp-emmy-enable-file-watchers)
+
+    (lsp)))
+
+(defun spacemacs//lua-setup-lsp-company ()
+  "Setup lsp auto-completion."
+  (spacemacs|add-company-backends
+    :backends company-lsp
+    :modes c-mode c++-mode
```
[the remark this answers is not in the document]
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: Hi @smile13241324 
You're totally right, it should for lua-mode. 
And I had update the PR to correct this error, https://github.com/syl20bnr/spacemacs/pull/12991/commits/2ea5c8f8c7997239a62ff10e2f4e60e35e01a4b3. 
Please review again. Thank you so much :)

**Comment by smile13241324** [comment]
Thank you for contributing to spacemacs :purple_heart:, I have squashed and cherry picked your commits into develop. I have also revised your commit and cleaned up the key bindings.

While I was doing so I have noticed that `lsp-ui-flycheck` is currently not called when LSP is activated instead the existing `Luacheck` library is used, however your original feature list said otherwise. If the intention was to have the LSP run the linter a new PR would be necessary.

See the code here which is the issue:
```elisp
(defun spacemacs//lua-setup-lsp-flycheck ()
  "Setup LSP Lua syntax checking."
  (when (spacemacs/enable-flycheck 'lua-mode)
    (lsp-ui-flycheck-enable nil)
    (flycheck-mode)))
```

**Comment by sunlin7** [comment]
Hi @smile13241324 , Thank you for your pasion. And the Luacheck based LSP is not ready in current commit. I'll try my best to make it happen if no one do that before me.

**Pull request closed without merging**